### PR TITLE
obfuscate: add exception when parsing empty-string identifiers

### LINKED ISSUE #316: obfuscate: "query not parsed" for a legitimate looking MySQL query
We're getting the `agent.parse.error` of `Query not parsed` for the following SQL query, which I think the SQL parser should be able to handle:

    /* ADB: /sample-path correlation_id:08a3cb269f3211e793b90e80c6e864fc */
    select id,field1,field2,field3 from OTable where evnt = %s 
    and OTable.status = %s
    and ((OTable.survey = "direct" and OTable.payment in ("scramble", "eggs")) 
        or (OTable.email is not null and OTable.email != "")) 
    and OTable.status != 350 order by %s

(I added newlines to the above to make it more readable - the query in question does not have newlines in it)

### PR DESCRIPTION
**Pull request by gbbr**
Fixes #316

### CODE REVIEW

File: obfuscate/sql_test.go [hunk 1]
```diff
@@ -340,6 +338,26 @@ FROM [Blogs] AS [b]
...
+			`SELECT * FROM users WHERE firstname=''`,
+			`SELECT * FROM users WHERE firstname = ?`,
+		},
+		{
+			`SELECT * FROM users WHERE firstname=' '`,
+			`SELECT * FROM users WHERE firstname = ?`,
+		},
+		{
+			`SELECT * FROM users WHERE firstname=""`,
+			`SELECT * FROM users WHERE firstname = ""`,
```
Comment (ufoot) [inline obfuscate/sql_test.go]: Sounds to me the problem is solved here, indeed.

File: obfuscate/sql_tokenizer.go [hunk 2]
```diff
@@ -408,7 +408,15 @@ func (tkn *Tokenizer) scanString(delim uint16, typ int) (int, []byte) {
 		}
 		buffer.WriteByte(byte(ch))
 	}
-	return typ, buffer.Bytes()
+	buf := buffer.Bytes()
+	if typ == ID && len(buf) == 0 || len(bytes.TrimSpace(buf)) == 0 {
+		// This string is an empty or white-space only identifier.
+		// We should keep the start and end delimiters in order to
+		// avoid creating invalid queries.
+		// See: https://github.com/DataDog/datadog-trace-agent/issues/316
```
Comment (ufoot) [inline obfuscate/sql_tokenizer.go]: 👍

**Comment by gbbr** [comment]
FYI Minor impact on time:
```
name                      old time/op    new time/op    delta
Tokenizer/Escaping/512-4    8.35µs ± 0%    8.47µs ± 0%  +1.53%
Tokenizer/Grouping/199-4    4.78µs ± 0%    4.92µs ± 0%  +2.82%

name                      old alloc/op   new alloc/op   delta
Tokenizer/Escaping/512-4    4.40kB ± 0%    4.40kB ± 0%   0.00%
Tokenizer/Grouping/199-4    2.72kB ± 0%    2.72kB ± 0%   0.00%

name                      old allocs/op  new allocs/op  delta
Tokenizer/Escaping/512-4      87.0 ± 0%      87.0 ± 0%   0.00%
Tokenizer/Grouping/199-4      68.0 ± 0%      68.0 ± 0%   0.00%
```

**Comment by gbbr** [comment]
Added another commit improving the white-space verification:
```
name                      old time/op    new time/op    delta
Tokenizer/Escaping/512-4    8.36µs ± 0%    8.37µs ± 0%  +0.13%
Tokenizer/Grouping/199-4    4.79µs ± 0%    4.71µs ± 0%  -1.69%

name                      old alloc/op   new alloc/op   delta
Tokenizer/Escaping/512-4    4.40kB ± 0%    4.40kB ± 0%   0.00%
Tokenizer/Grouping/199-4    2.72kB ± 0%    2.72kB ± 0%   0.00%

name                      old allocs/op  new allocs/op  delta
Tokenizer/Escaping/512-4      87.0 ± 0%      87.0 ± 0%   0.00%
Tokenizer/Grouping/199-4      68.0 ± 0%      68.0 ± 0%   0.00%
```

**Pull request merged**